system/security/pkencrypt/pkencrypt.cpp

```diff
@@ -0,0 +1,260 @@
+#endif
+    }
+
+    virtual ~_CPKEncryptionManager()
+    {
+#ifdef _USE_ZLIB
+        {
+            CriticalBlock b(encrypterCrit);
+            if (pEncrypter)
+                releaseIZ(pEncrypter);
+        }
+
+        {
+            CriticalBlock b(decrypterCrit);
+            if (pDecrypter)
+                releaseIZ(pDecrypter);
+        }
+#endif
+    }
+
+    virtual inline bool isEncrypterConfigured()
+    {
+        return encrpytConfigured;
+    }
+
+    virtual bool PKEncrypt(const char * _str, StringAttr & _key, StringAttr & _token)
+    {
+        if (!encrpytConfigured)
+            throw MakeStringException(-1, "PKEncrypt error: encryption not configured");
+
+        if (_str == nullptr  || *_str == '\0')
+            return false;
+
+#ifdef _USE_ZLIB
+        CriticalBlock b(encrypterCrit);
+
+        if (!pEncrypter)
+        {
+            StringBuffer pubKey;
+            pubKey.loadFile(publicKeyFile);//throws on error
+            pEncrypter = createZEncryptor(pubKey.str());
+            if (pEncrypter)
+            {
+                pEncrypter->setEncoding(true);
+#ifdef _DEBUG
+                pEncrypter->setTraceLevel(100);
+#endif
+            }
+            else
+                throw MakeStringException(-1, "PKEncrypt error: creating PKI encrypter using key file %s", publicKeyFile.str());
+        }
+
+        IZBuffer* buf = nullptr;
+        IZBuffer* key = nullptr;
+        int rc = pEncrypter->encrypt(strlen(_str), (unsigned char*)_str, key, buf);
+        if (rc == 0)
+        {
+            if (buf->length())
+                _token.set((const char *)buf->buffer(), buf->length());
+            if (key->length())
+                _key.set((const char *)key->buffer(), key->length());
+            releaseIZ(buf);
+            releaseIZ(key);
+            return true;//success
+        }
+        else
+            throw MakeStringException(-1, "PKEncrypt error: %d PKI encrypting token using key file %s", rc, publicKeyFile.str());
+#else
+        throw MakeStringException(-1, "PKEncrypt error: Platform built without ZLIB!");
+#endif
+        return false;//fail
+    }
+
+    virtual inline bool isDecrypterConfigured()
+    {
+        return decrpytConfigured;
+    }
+
+    virtual bool PKDecrypt(const char * _key, const char * _token, StringAttr & _result)
+    {
+        if (!decrpytConfigured)
+            throw MakeStringException(-1, "PKEncrypt error: decryption not enabled");
+
+        if (_key == nullptr || *_key == '\0')
+            return false;
+
+#ifdef _USE_ZLIB
+        CriticalBlock b(decrypterCrit);
+
+        if (!pDecrypter)
+        {
+            StringBuffer privateKeyBuff;
+            privateKeyBuff.loadFile(privateKeyFile.str());
+            if (privateKeyBuff.isEmpty())
+                throw MakeStringException(-1, "PKEncrypt error: Could not read private key file");
+            pDecrypter = createZDecryptor(privateKeyBuff.str(), nullptr);//create once
```

Comment:
It may not be used often or at all, but we should probably support a passphrase.
At the least the creator of the private key should be aware that a passphrase is not supported here.

Reply:
Added support for _passphrase